Pressure inlet boundary condition in the Navier Stokes FV action

snschune opened the pull request:
closes #21380 

## Reason

Chimney like natural convection problems require pressure inlet conditions. Other problems
might need that too.

## Design

Small changes to the NSFV action. 

## Impact

The input for the momentum inlet function had to change slightly
to accommodate pressure functions.

@grmnptr

Commit:
Pressure inlet boundary condition in the finite volume action (#21380)

lindsayad commented on modules/navier_stokes/src/actions/NSFVAction.C:
```diff
@@ -2370,20 +2392,45 @@ NSFVAction::checkBoundaryParameterErrors()
+                 "momentum_inlet_types");
+
+  unsigned int k = 0;
+  for (unsigned int j = 0; j < _momentum_inlet_types.size(); ++j)
+  {
+    const auto & type = _momentum_inlet_types[j];
```
```suggestion
  for (const auto & type : _momentum_inlet_types)
  {
```

lindsayad commented on modules/navier_stokes/src/actions/NSFVAction.C:
```diff
@@ -2370,20 +2392,45 @@ NSFVAction::checkBoundaryParameterErrors()
+                 "fixed-velocity and fixed-pressure entries",
+                 "momentum_inlet_types");
+
+  unsigned int k = 0;
```
```suggestion
  // index into _momentum_inlet_function
  unsigned int k = 0;
```

lindsayad commented on modules/navier_stokes/test/tests/finite_volume/ins/action/errors/tests:
```diff
@@ -12,9 +12,17 @@
+  [momentum-inlet-function-error-2]
+    type = 'RunException'
+    input = 2d-rc-error-action.i
+    requirement = 'The system shall throw an error if the number of entries in momentum_inlet_types does not match the total number of fixed-velocity and fixed-pressure inlet boundaries.'
```
We've been told that requirement language should be as free of implementation details as possible. I think here you can just do
```suggestion
    requirement = 'The system shall throw an error if the number of entries for momentum inlet types does not match the total number of fixed-velocity and fixed-pressure inlet boundaries.'
```

lindsayad reviewed (changes requested):
I recently tried a pairing of inlet and outlet pressure boundary conditions and failed, so it's nice to see that you were able to drum up a test case

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1088047/) on 7b6d1bc wanted to post the following:


## Framework coverage

Coverage did not change


## Modules coverage

### Navier stokes

<table>
  <thead>
    <tr>
      <th></th>
      <th>41a07d</th>
      <th colspan=4>#21406 7b6d1b</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.57%</td>
      <td>85.59%</td>
      <td>+0.03%</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>9723</td>
      <td>9744</td>
      <td>+21</td>
      <td>26</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>1640</td>
      <td>1640</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/21406/coverage_diff/navier_stokes)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/21406/coverage/navier_stokes)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/21406/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/21406/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/21406/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/21406/coverage/contact)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/21406/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/21406/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/21406/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/21406/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/21406/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/21406/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/21406/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/21406/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/21406/coverage/navier_stokes)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/21406/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/21406/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/21406/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/21406/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/21406/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/21406/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/21406/coverage/richards)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/21406/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/21406/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/21406/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/21406/coverage/xfem)

</details>



This comment will be updated on new commits.

snschune commented:
@lindsayad I made that test a verification exercise and it converges to the hand calculation.

GiudGiud commented on modules/navier_stokes/src/actions/NSFVAction.C:
```diff
@@ -2370,20 +2392,45 @@ NSFVAction::checkBoundaryParameterErrors()
+                     "momentum_inlet_function",
+                     _dim,
+                     "entries ",
+                     " the momentum_inlet_types subvector for fixed-velocity");
```
```suggestion
                     " the momentum_inlet_types subvector for fixed-velocity on inlet: " + _inlet_boundaries[k]);
```

GiudGiud commented on modules/navier_stokes/src/actions/NSFVAction.C:
```diff
@@ -2370,20 +2392,45 @@ NSFVAction::checkBoundaryParameterErrors()
+                     "momentum_inlet_function",
+                     1,
+                     "entries ",
+                     " the momentum_inlet_types subvector for fixed-pressure");
```
```suggestion
                     " the momentum_inlet_types subvector for fixed-pressure inlet : " + _inlet_boundaries[k]);
```

GiudGiud committed:
Apply suggestions from code review
Make error message more explicit about which inlet is concerned
Add size explicitly to error message to avoid user thinking we re mentioning the boundary id

Co-authored-by: Alex Lindsay <alexlindsay239@gmail.com>

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1089353/) on da1c85a wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/21406/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/21406/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 41a07d673011fdbe59b4b71450ab668c678c716e`

GiudGiud commented:
Seems the new test is diffing in parallel on integral quantities

grmnptr committed:
Modify tolerances and scaling to pass tests in parallel. (#21380)

grmnptr commented:
Okay, turned out to be a tolerance issue. Should be good now.
